internal/exposer/exposer.go

```diff
@@ -0,0 +1,488 @@
+			return nil, fmt.Errorf("failed to parse bindCIDR: %w", err)
+		}
+
+		bindCIDRPrefixes = append(bindCIDRPrefixes, prefix)
+	}
+
+	return &Exposer{
+		AnnotationKey:     annotationKey,
+		BindCIDRPrefixes:  bindCIDRPrefixes,
+		manager:           mgr,
+		logger:            logger,
+		hostPortToMapping: map[int]*portMapping{},
+		svcToLB:           map[string]*loadBalancer{},
+	}, nil
+}
+
+type portMapping struct {
+	hostIPSet map[string]struct{}
+	svcName   string
+	svcPort   int
+}
+
+type loadBalancer struct {
+	lb       loadbalancer.TCP
+	hostPort int
+}
+
+// Exposer is a controller that exposes the given services on the given host interfaces.
+type Exposer struct {
+	manager           manager.Manager
+	hostPortToMapping map[int]*portMapping
+	svcToLB           map[string]*loadBalancer
+	logger            *zap.Logger
+	ipSet             map[string]struct{}
+	AnnotationKey     string
+	BindCIDRPrefixes  []netip.Prefix
+	mappingLock       sync.Mutex
+	ipSetLock         sync.Mutex
+}
+
+// Run runs the Exposer.
+func (e *Exposer) Run(ctx context.Context) error {
+	e.logger.Info(
+		"starting Exposer",
+		zap.String("annotationKey", e.AnnotationKey),
+		zap.Strings("bindCIDRs", slices.Map(e.BindCIDRPrefixes, func(prefix netip.Prefix) string {
+			return prefix.String()
+		})),
+	)
+
+	if len(e.BindCIDRPrefixes) == 0 {
+		e.logger.Info("bindCIDRs are empty, mappings will listen on all interfaces")
+	}
+
+	ctrller, err := controller.New(version.Name+"-controller", e.manager,
+		controller.Options{
+			Reconciler:         e,
+			NeedLeaderElection: pointer.Bool(false),
+		})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	if err = ctrller.Watch(source.Kind(e.manager.GetCache(), &corev1.Service{}), &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to watch Services: %w", err)
+	}
+
+	var eg errgroup.Group
+
+	eg.Go(func() error {
+		if err = e.manager.Start(ctx); err != nil {
+			return fmt.Errorf("failed to start manager: %w", err)
+		}
+
+		return nil
+	})
+
+	if len(e.BindCIDRPrefixes) > 0 {
+		e.logger.Info("bindCIDRs are specified, start IP change listener")
+
+		eg.Go(func() error {
+			if err = e.runIPChangeListener(ctx); err != nil {
+				return fmt.Errorf("failed to start IP change listener: %w", err)
+			}
+
+			return nil
+		})
+	} else {
+		e.logger.Info("bindCIDRs are empty, IP change listener will not be started")
+	}
+
+	return eg.Wait()
+}
+
+// runIPChangeListener starts a goroutine that listens for changes on the IPs on the network interfaces of the host.
+// If there is a change since the last check, the IP set is refreshed and all services a re-reconciled.
+func (e *Exposer) runIPChangeListener(ctx context.Context) error {
+	ticker := time.NewTicker(30 * time.Second)
+	defer ticker.Stop()
+
+	logger := e.logger.With(zap.String("component", "ip-change-listener"))
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+		}
```

Comment:
```suggestion
		if _, ok := channel.RecvWithContext(ctx, ticker.C); !ok && ctx.Err() != nil {
			return nil
		}
```

https://github.com/siderolabs/gen/blob/main/channel/recv.go#L12

Reply:
I think `channel.RecvWithContext` is a bit confusing, I'd prefer to keep it the way it is tbh.